src/wallet/sqlite.cpp

```diff
@@ -0,0 +1,629 @@
+    sqlite3_reset(m_read_stmt);
+    return ret;
+}
+
+bool SQLiteBatch::StartCursor()
+{
+    assert(!m_cursor_init);
+    if (!m_database.m_db) return false;
+    m_cursor_init = true;
+    return true;
+}
+
+bool SQLiteBatch::ReadAtCursor(CDataStream& key, CDataStream& value, bool& complete)
+{
+    complete = false;
+
+    if (!m_cursor_init) return false;
+
+    int res = sqlite3_step(m_cursor_stmt);
+    if (res == SQLITE_DONE) {
+        complete = true;
+        return true;
+    }
+    if (res != SQLITE_ROW) {
+        LogPrintf("SQLiteBatch::ReadAtCursor: Unable to execute cursor step: %s\n", sqlite3_errstr(res));
+        return false;
+    }
+
+    // Leftmost column in result is index 0
+    const char* key_data = reinterpret_cast<const char*>(sqlite3_column_blob(m_cursor_stmt, 0));
+    int key_data_size = sqlite3_column_bytes(m_cursor_stmt, 0);
+    key.write(key_data, key_data_size);
+    const char* value_data = reinterpret_cast<const char*>(sqlite3_column_blob(m_cursor_stmt, 1));
+    int value_data_size = sqlite3_column_bytes(m_cursor_stmt, 1);
+    value.write(value_data, value_data_size);
+    return true;
+}
+
+void SQLiteBatch::CloseCursor()
+{
+    sqlite3_reset(m_cursor_stmt);
+    m_cursor_init = false;
+}
+
+bool SQLiteBatch::TxnBegin()
+{
+    if (!m_database.m_db || sqlite3_get_autocommit(m_database.m_db) == 0) return false;
+    int res = sqlite3_exec(m_database.m_db, "BEGIN TRANSACTION", nullptr, nullptr, nullptr);
+    if (res != SQLITE_OK) {
+        LogPrintf("SQLiteBatch: Failed to begin the transaction\n");
+    }
+    return res == SQLITE_OK;
+}
+
+bool SQLiteBatch::TxnCommit()
+{
+    if (!m_database.m_db || sqlite3_get_autocommit(m_database.m_db) != 0) return false;
+    int res = sqlite3_exec(m_database.m_db, "COMMIT TRANSACTION", nullptr, nullptr, nullptr);
+    if (res != SQLITE_OK) {
+        LogPrintf("SQLiteBatch: Failed to commit the transaction\n");
+    }
+    return res == SQLITE_OK;
+}
+
+bool SQLiteBatch::TxnAbort()
+{
+    if (!m_database.m_db || sqlite3_get_autocommit(m_database.m_db) != 0) return false;
+    int res = sqlite3_exec(m_database.m_db, "ROLLBACK TRANSACTION", nullptr, nullptr, nullptr);
+    if (res != SQLITE_OK) {
+        LogPrintf("SQLiteBatch: Failed to abort the transaction\n");
+    }
+    return res == SQLITE_OK;
+}
+
+bool ExistsSQLiteDatabase(const fs::path& path)
+{
+    const fs::path file = path / DATABASE_FILENAME;
+    return fs::symlink_status(file).type() == fs::regular_file && IsSQLiteFile(file);
+}
+
+std::unique_ptr<SQLiteDatabase> MakeSQLiteDatabase(const fs::path& path, const DatabaseOptions& options, DatabaseStatus& status, bilingual_str& error)
+{
+    const fs::path file = path / DATABASE_FILENAME;
+    try {
+        auto db = MakeUnique<SQLiteDatabase>(path, file);
+        if (options.verify && !db->Verify(error)) {
+            status = DatabaseStatus::FAILED_VERIFY;
+            return nullptr;
+        }
+        return db;
```

Comment:
Does this return path need to set `status = DatabaseStatus::SUCCESS;`?

I am asking because gcc seems to generate binaries that will inject false positives into valgrind. I haven't found a bug in the code.

Though, the bdb version also seems to set `::SUCCESS`, so it might be appropriate here as well?

Reply:
I suppose it should, but istm it shouldn't matter. I'm pretty sure we never check `status` if there isn't a failure.

Reply:
Jup, we don't check `status` if there is no failure. It is just a style question.

Reply:
Also happening with clang, so I'll try to fix this

Reply:
#20324